### PR TITLE
LibJS: Start implementing Intl.Segmenter

### PR DESCRIPTION
**Pull request by IdanHo**
This implements essentially everything but the Segmentation itself, which will happen in a future PR (hopefully together with the https://github.com/tc39/ecma402/pull/653 related changes).

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/Intl/Segments.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    virtual ~Segments() override = default;
+
+    Segmenter& segments_segmenter() const { return m_segments_segmenter; }
+
+    String const& segments_string() const { return m_segments_string; }
+
+private:
+    virtual void visit_edges(Cell::Visitor&) override;
+
+    Segmenter& m_segments_segmenter; // [[SegmentsSegmenter]]
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Intl/Segments.h:31]: :^)

File: Userland/Libraries/LibJS/Runtime/Intl/SegmentsPrototype.cpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+// 18.5.2 The %SegmentsPrototype% Object, https://tc39.es/ecma402/#sec-%segmentsprototype%-object
+SegmentsPrototype::SegmentsPrototype(GlobalObject& global_object)
+    : PrototypeObject(*global_object.object_prototype())
+{
+}
+
+void SegmentsPrototype::initialize(GlobalObject& global_object)
+{
+    Object::initialize(global_object);
+}
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Intl/SegmentsPrototype.cpp:45]: Do we even need to override this if all we do is calling the base `initialize()`? Not sure how we do this elsewhere.

File: Userland/Libraries/LibJS/Runtime/Intl/SegmentIteratorPrototype.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+void SegmentIteratorPrototype::initialize(GlobalObject& global_object)
+{
+    Object::initialize(global_object);
+
+    auto& vm = this->vm();
+
+    // 18.6.2.2 %SegmentIteratorPrototype% [ @@toStringTag ], https://tc39.es/ecma402/#sec-%segmentiteratorprototype%.@@tostringtag
+    define_direct_property(*vm.well_known_symbol_to_string_tag(), js_string(global_object.heap(), "Segmenter String Iterator"), Attribute::Configurable);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Intl/SegmentIteratorPrototype.cpp:26]: Can we add a test for this? :)

File: Userland/Libraries/LibJS/Runtime/Intl/SegmentsPrototype.cpp [hunk 4]
```diff
@@ -19,6 +20,28 @@ SegmentsPrototype::SegmentsPrototype(GlobalObject& global_object)
 void SegmentsPrototype::initialize(GlobalObject& global_object)
 {
     Object::initialize(global_object);
+
+    auto& vm = this->vm();
+
+    u8 attr = Attribute::Writable | Attribute::Configurable;
+    define_native_function(*vm.well_known_symbol_iterator(), symbol_iterator, 0, attr);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Intl/SegmentsPrototype.cpp:27]: Can we add a test for this? :)

File: Userland/Libraries/LibJS/Runtime/Intl/Segmenter.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        Grapheme,
+        Word,
+        Sentence,
+    };
+
+    explicit Segmenter(Object& prototype);
+    virtual ~Segmenter() override = default;
+
+    String const& locale() const { return m_locale; }
+    void set_locale(String locale) { m_locale = move(locale); }
```
Comment (trflynn89) [inline Userland/Libraries/LibJS/Runtime/Intl/Segmenter.h:28]: Not really a review comment, but you might find you need to store the `data_locale` from `ResolveLocale` as well. That string is similar, but has Unicode extensions removed, which provides better LibUnicode lookups. It's been important on some Intl objects, so just something to keep in mind if you hit locale specific issues.

File: Userland/Libraries/LibJS/Runtime/Intl/SegmenterConstructor.cpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    // 9. Let localeData be %Segmenter%.[[LocaleData]].
+
+    // 10. Let r be ResolveLocale(%Segmenter%.[[AvailableLocales]], requestedLocales, opt, %Segmenter%.[[RelevantExtensionKeys]], localeData).
+    auto result = resolve_locale(requested_locales, opt, {});
+
+    // 11. Set segmenter.[[Locale]] to r.[[locale]].
+    segmenter->set_locale(move(result.locale));
+
+    // 12. Let granularity be ? GetOption(options, "granularity", "string", « "grapheme", "word", "sentence" », "grapheme").
+    auto granularity = TRY(get_option(global_object, *options, vm.names.granularity, Value::Type::String, { "grapheme"sv, "word"sv, "sentence"sv }, "grapheme"sv));
```
Comment (trflynn89) [inline Userland/Libraries/LibJS/Runtime/Intl/SegmenterConstructor.cpp:81]: Oh good, they picked the scariest one as the default :laughing:

File: Userland/Libraries/LibJS/Tests/builtins/Intl/Segmenter/Segmenter.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+    test("called without new", () => {
+        expect(() => {
+            Intl.Segmenter();
+        }).toThrowWithMessage(TypeError, "Intl.Segmenter constructor must be called with 'new'");
+    });
+});
+
+describe("normal behavior", () => {
+    test("length is 0", () => {
+        expect(Intl.Segmenter).toHaveLength(0);
```
Comment (trflynn89) [inline Userland/Libraries/LibJS/Tests/builtins/Intl/Segmenter/Segmenter.js:11]: These tests are a bit lacking given this commit also implements the full constructor. Generally the Intl constructor tests try to hit every exceptional case in the "errors" section, and every valid option in the "normal behavior" section. Mostly a copy-paste job from existing Intl tests.

File: Userland/Libraries/LibJS/Runtime/Intl/Segments.h [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace JS::Intl {
+
+class Segments final : public Object {
+    JS_OBJECT(Segments, Object);
+
+public:
+    static Segments* create(GlobalObject&, Segmenter&, String);
+
+    explicit Segments(GlobalObject&, Segmenter&, String);
```
Comment (trflynn89) [inline Userland/Libraries/LibJS/Runtime/Intl/Segments.h:21]: `explicit` not needed here

File: Userland/Libraries/LibJS/Runtime/Intl/SegmentsPrototype.cpp (hunk 2 above)
Comment (trflynn89) [reply Userland/Libraries/LibJS/Runtime/Intl/SegmentsPrototype.cpp:45]: This becomes more filled in a couple of commits later - `LibJS: Implement Intl %SegmentsPrototype%[@@iterator] `

File: Userland/Libraries/LibJS/Runtime/Intl/Segments.h (hunk 8 above)
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/Intl/Segments.h:21]: fixed

File: Userland/Libraries/LibJS/Runtime/Intl/SegmentIteratorPrototype.cpp (hunk 3 above)
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/Intl/SegmentIteratorPrototype.cpp:26]: Added

File: Userland/Libraries/LibJS/Runtime/Intl/SegmentsPrototype.cpp (hunk 4 above)
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/Intl/SegmentsPrototype.cpp:27]: Added

File: Userland/Libraries/LibJS/Tests/builtins/Intl/Segmenter/Segmenter.js (hunk 7 above)
Comment (IdanHo) [reply Userland/Libraries/LibJS/Tests/builtins/Intl/Segmenter/Segmenter.js:47]: Fixed

**Review by linusg — APPROVED** [review]

**Comment by trflynn89** [comment]
"trflynn89 self-requested a review now"
...I don't even know what button I pressed to do that, but ignore that :laughing:

**Comment by linusg** [comment]
Ref #9633

**Pull request merged**